proposals/pattern-const-expressions.md

```diff
@@ -0,0 +1,145 @@
+
+## Detailed design
+[design]: #detailed-design
+
+A pattern expression is constant if all its contained subpatterns are all also constant. The following subpatterns are constant:
+- Boolean literal (`true` or `false`)
+- Numeric literal (e.g. `0`, `1f`, `3m`, etc.)
+- Character literal (e.g. `'c'`, `'\0'`, etc.)
+- String literal (e.g. `"Hello world."`, `""`, etc.)
+- Relative numeric literal (e.g. `> 3`, `< 0.3m`, etc.)
+- Relative character literal (e.g. `> 'a'`, `<= 'z'`, etc.)
+- Null literal (`null`)
+- Default literal (`default(T)`), but not `default` without a type
+- And, or and not patterns (e.g. the *and* in  `>= 'a' and <= 'z'`)
+- Parenthesized patterns (e.g. the parentheses in `is (not (X or Y))`)
+- Constant field (including enum fields) or local, e.g.
+  - `x is y` when y is a const field or local, or
+  - `x is SomeEnum.None` when `None` is an enum field
+
+The allowed subpatterns as shown in the list above are called "constant subpatterns", as they will be eligible for compile-time evaluation.
+
+If the `is` expression consists of a constant expression on the LHS, and is compared against a constant pattern on the right, the entire `is` expression is a constant expression.
+
+### Spec
+[spec]: #spec
+
+The [§12.23 section](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/expressions.md#1223-constant-expressions) in the spec needs to be updated accordingly, in the following segments:
+
+> Only the following constructs are permitted in constant expressions:
+> - Literals (including the null literal).
+> - References to const members of class and struct types.
+> - References to members of enumeration types.
+> - References to local constants.
+> - Parenthesized subexpressions, which are themselves constant expressions.
+> - Cast expressions.
+> - checked and unchecked expressions.
+> - nameof expressions.
+> - The predefined +, –, !, and ~ unary operators.
+> - The predefined +, –, *, /, %, <<, >>, &, |, ^, &&, ||, ==, !=, <, >, <=, and >= binary operators.
+> - The ?: conditional operator.
+> - sizeof expressions, provided the unmanaged-type is one of the types specified in §23.6.9 for which sizeof returns a constant value.
+> - Default value expressions, provided the type is one of the types listed above.
+> - **`is` expressions with only the following subpatterns:**
+>   - **Boolean literal patterns.**
+>   - **Numeric literal patterns.**
+>   - **Character literal patterns.**
+>   - **String literal patterns.**
+>   - **Relative numeric literal patterns.**
+>   - **Relative character literal patterns.**
+>   - **Null literal patterns.**
+>   - **Default literal patterns.**
+>   - **And, or and not patterns.**
+>   - **Parenthesized patterns.**
+>   - **References to constant fields or locals inside constant patterns.**
```

Comment:
Can we remove the word `literal` from these? It doesn't seem to be necessary, given the recursive nature of "constant expressions must be composed of other constant expressions", and it will reduce the brittleness here for future changes.